equinix/resource_metal_device_acc_test.go

```diff
@@ -400,6 +400,36 @@ func TestAccMetalDevice_IPXEConfigMissing(t *testing.T) {
 	})
 }
 
+func TestAccMetalDevice_allowUserdataChanges(t *testing.T) {
+	var d1, d2 packngo.Device
+	rs := acctest.RandString(10)
+	rInt := acctest.RandInt()
+	r := "equinix_metal_device.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccMetalDeviceCheckDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccMetalDeviceConfig_allowUserdataChanges(rInt, rs),
+				Check: resource.ComposeTestCheckFunc(
+					testAccMetalDeviceExists(r, &d1),
+					resource.TestCheckResourceAttr(r, "user_data", fmt.Sprintf("#!/usr/bin/env sh\necho 'Allow userdata changes %d'\n", rInt)),
+				),
+			},
+			{
+				Config: testAccMetalDeviceConfig_allowUserdataChanges(rInt+1, rs),
+				Check: resource.ComposeTestCheckFunc(
```

Comment:
I think we should be able to test that the device would have been destroyed rather than waiting for the full destroy/create cycle.

https://github.com/hashicorp/terraform-provider-google/issues/12428#issuecomment-1232142860 for example would cause the step to fail with an error that we could compare. 

Perhaps `PlanOnly` would help with that too.

Reply:
The comment you linked seems to be an unimplemented proposal to address an issue with detecting when resources are recreated in tests; I didn't see anything actionable in there.

Since this is an acceptance test, I think we benefit from actually applying the plan here: this test confirms both that the device was updated in-place and that the attribute value was updated correctly.

Reply:
Apparently, I was reading more into it than what the commenter posted.  

```hcl
{
Config: `resource "equinix_metal_device" "foo" {
   behavior { allow_changes = false }
   userdata = "original"  ...
   lifecycle { prevent_destroy = true }
}`
PlanOnly: true
},
{
Config: `resource "equinix_metal_device" "foo" {
   behavior { allow_changes = false }
   userdata = "new"  ...
   lifecycle { prevent_destroy = true }
}`
PlanOnly: true,
ExpectError: regexp.MustCompile(`something about prevent_destroy`),
},
{
Config: `resource "equinix_metal_device" "foo" {
   behavior { allow_changes = false }
   userdata = "new"  ...
}`
PlanOnly: true,
},
```